pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (rw *revisionWatcher) runOnce(tickCh <-chan time.Time) bool {
+	select {
+	case dests := <-rw.destsChan:
+		if dests == nil {
+			// shutdown
+			return true
+		}
+		rw.dests = dests
+		rw.checkDests()
+	case <-tickCh:
+		rw.checkDests()
+	}
+	return false
+}
+
+func (rw *revisionWatcher) run() {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if rw.runOnce(ticker.C) {
+			return
+		}
+	}
+}
+
+// RevisionBackendsManager listens to revision endpoints and keeps track of healthy
+// l4 dests which can be used to reach a revisoin
+type RevisionBackendsManager struct {
+	revisionWatchers    map[RevisionID]*revisionWatcher
+	revisionDestsCh     map[RevisionID]chan []string
+	revisionWatchersMux sync.RWMutex
+
+	updateCb  RevisionDestsUpdateFunc
+	transport http.RoundTripper
+	logger    *zap.SugaredLogger
+}
+
+// NewRevisianBackendsManager returnes a RevisionBackendsManager
+func NewRevisionBackendsManager(updateCb RevisionDestsUpdateFunc, transport http.RoundTripper,
+	endpointsInformer corev1informers.EndpointsInformer, logger *zap.SugaredLogger) *RevisionBackendsManager {
+	rbm := &RevisionBackendsManager{
+		revisionWatchers: make(map[RevisionID]*revisionWatcher),
+		revisionDestsCh:  make(map[RevisionID]chan []string),
+		updateCb:         updateCb,
+		transport:        transport,
+		logger:           logger,
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(serving.RevisionUID),
+			// We are only interested in the private services, since that is
+			// what is populated by the actual revision backends.
+			reconciler.LabelFilterFunc(networking.ServiceTypeKey, string(networking.ServiceTypePrivate), true),
+		),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    rbm.endpointsUpdated,
+			UpdateFunc: controller.PassNew(rbm.endpointsUpdated),
+			DeleteFunc: rbm.endpointsDeleted,
+		},
+	})
+
+	return rbm
+}
+
+func (rbm *RevisionBackendsManager) getOrCreateDestsCh(rev RevisionID) chan []string {
+	destsCh, ok := rbm.revisionDestsCh[rev]
+	if !ok {
+		destsCh = make(chan []string)
+		rw := newRevisionWatcher(rev, rbm.updateCb, destsCh, rbm.transport, rbm.logger)
+		rbm.revisionDestsCh[rev] = destsCh
+		rbm.revisionWatchers[rev] = rw
+		go rw.run()
+	}
+
+	return destsCh
+}
+
+// endpointsUpdated is a handler function to be used by the Endpoints informer.
+// It updates the endpoints in the RevisionBackendsManager if the hosts changed
+//
+// This function must not be called in parallel to not induce a wrong order of events.
+func (rbm *RevisionBackendsManager) endpointsUpdated(newObj interface{}) {
+	endpoints := newObj.(*corev1.Endpoints)
+	revID := RevisionID{endpoints.Namespace, endpoints.Labels[serving.RevisionLabelKey]}
+	rbm.logger.Infof("rev %v endpoints updated", revID)
+
+	rbm.revisionWatchersMux.Lock()
+	defer rbm.revisionWatchersMux.Unlock()
```

Comment:
Can we move the locking into the mutating functions themselves (getOrCreate + deleteRevisionWatcher)? It seems to not affect correctness in this case and it's imo preferable to acquire locks in the same function where the actual mutations are done instead of requiring locks to be held when calling a function.

Reply:
I think I can for get, for delete its a perf/code improvement where i'd need to make a separate implementation for Clear (which loops over all keys not just the one).